### PR TITLE
Call default_factory during `model_construct` (finish fixing test_construction.py)

### PR DESCRIPTION
**Pull request by dmontagu**
I'll just note that `tests.test_main.test_default_factory_called_once` does show a dubious but understandable use case for wanting the `get_default` function to be called _only_ when instantiating a model. (And that test fails if you do the trivial thing and just change `get_default` to always call `default_factory()`.)

However, it was easy enough to just add (what amounts to) a boolean flag argument for this, so I think this is a good solution.

Happy to rename arguments or whatever if preferred.

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by dmontagu** [comment]
please review

File: pydantic/fields.py [hunk 1]
```diff
@@ -243,10 +243,16 @@ def _collect_metadata(cls, kwargs: dict[str, Any]) -> list[Any]:
             metadata.append(_fields.PydanticGeneralMetadata(**general_metadata))
         return metadata
 
-    def get_default(self) -> Any:
-        # we don't want to call default_factory as it may have side effects, so we default to None as the
-        # least-worse alternative
-        return _utils.smart_deepcopy(self.default) if self.default_factory is None else None
+    def get_default(self, force_call_default_factory: bool = False) -> Any:
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/fields.py:246]: also `get_default` isn't used that much, seems like the only other place is in building signatures.

I'd therefore be open to making `call_default_factory` required.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

**Comment by samuelcolvin** [comment]
please update.

**Comment by dmontagu** [comment]
please review (and merge if it looks good)

File: pydantic/fields.py [hunk 2]
```diff
@@ -243,10 +243,16 @@ def _collect_metadata(cls, kwargs: dict[str, Any]) -> list[Any]:
...
+        # we generally don't want to call default_factory as it may have side effects,
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/fields.py]: done, also updated the contents to reflect the fact the argument is now required

File: pydantic/fields.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/fields.py]: done

**Comment by dmontagu** [comment]
Given feedback has been addressed, I'll merge this now. But just comment here if you want other changes and I'll open a new PR making them.

**Pull request merged**